### PR TITLE
Battle effects screen

### PR DESCRIPTION
**Pull request by jtigues**
- Adds a new screen to the Tracker during trainer battles. Accessed by clicking on the pokeball icons when viewing an enemy mon
- The screen can show Statuses of individual mons, as well as effects applied to either side of the battle, and the whole battle. User can switch between the different views by 
- The screen allows users to see passive Battle Effects which the game has notified them of, but cannot otherwise be reminded of once the next turn starts.
- Some examples: Weather conditions, Leech Seed and which mon will receive the HP, Taunt, Encore, Confusion, Reflect/Light Screen, pending Future Sight attacks, or the number of times an enemy has successfully Protected consecutively.
image

There is one known and incredibly minor bug I have not been able to sort out: I cannot for the life of me get the text & buttons to appear with shadows, and it seems it should work fine enough. But functionally, everything works fine across Games and Languages.
Another note: the plan is eventually to have the Last Move used by each pokemon displayed separately here. But that framework is not in place yet, so it just shows the Last Move used as recorded by the Last Damage framework
Also, I don't know where this expYield removal change came from, if I need to go in and manually re-add all of those that's fine.

### CODE REVIEW

**Comment by jtigues** [comment]
Few small things I've noticed looking back over it:
- "stale" flag is no longer used but still has references
- BattleEffectsScreen Button is being drawn unnecessarily, the button is invisible
- Left Arrow pixel image has an extra column of 0s, it's actually 10 by 9.

**Review by Fellshadow — CHANGES_REQUESTED** [review]
the removal of the `ExpYield` in the PR is really weird to me
On the PR itself it looks like [this commit](https://github.com/besteon/Ironmon-Tracker/pull/428/commits/7c546b601acd10da9b0c02b1a07d647ac806ddd4) is the one github thinks is removing it, but weirdly the same commit doesn't look like that at all on the actual branch's commit history (7c546b601acd10da9b0c02b1a07d647ac806ddd4)?

What's even weirder is that the merge commit from Zac merging in the PR is present on the branch's commits *but* it doesn't seem to be there when viewing the [files on the branch](<https://github.com/jtigues/Ironmon-Tracker/tree/BattleEffectsScreen/ironmon_tracker/Languages>)?

I think the commit history got messed up somewhere somehow 🤔 idk if a rebase would work? Rest of the code seems fine though (Aside from what you've mentioned and the one comment below) so manually adding back in also would be fine i think, though commit history may/may not be messy

File: ironmon_tracker/screens/BattleEffectsScreen.lua [hunk 1]
```diff
@@ -0,0 +1,1052 @@
...
+		#define BATTLE_TERRAIN_LONG_GRASS   1
+		#define BATTLE_TERRAIN_SAND         2
+		#define BATTLE_TERRAIN_UNDERWATER   3
+		#define BATTLE_TERRAIN_WATER        4
+		#define BATTLE_TERRAIN_POND         5
+		#define BATTLE_TERRAIN_MOUNTAIN     6
+		#define BATTLE_TERRAIN_CAVE         7
+		#define BATTLE_TERRAIN_BUILDING     8
+	]]
+	local battleTerrain = Memory.readbyte(0x02022ff0)
```
[the remark this answers is not in the document]
Comment (jtigues) [reply ironmon_tracker/screens/BattleEffectsScreen.lua]: Fixed, ty

**Comment by jtigues** [comment]
> Few small things I've noticed looking back over it:
> 
> * "stale" flag is no longer used but still has references
> * BattleEffectsScreen Button is being drawn unnecessarily, the button is invisible
> * Left Arrow pixel image has an extra column of 0s, it's actually 10 by 9.

Fixed

**Review by UTDZac — CHANGES_REQUESTED** [review]
I quickly reviewed the files that weren't the large `BattleEffectsScreen.lua` file, I'll review that one later.

As a note for that file though, I noticed quite a few functions that were defined as global, such as `function loadFieldEffects()`. If this function is only meant to be used within this file, use `local function loadFieldEffects()`, otherwise, preface it with the filename: `function BattleEffectsScreen.loadFieldEffects()`.

I also recommend making two changes to the `initialize() function:
- Move this as far up to the top of the file as possible, its usually the first important thing that should be seen for any given screen file
- Include any `BattleEffectsScreen.variablename` that have defined defaults that can change in this function. Quickloading does not reset those variables, so you must reset them in this function (called when the tracker starts up).
   - For example: The variable `viewedMonIndex` is one that I'd expect to change as I play the game. It needs to be set to its default value in `initialize()`, otherwise if I reset my game while its some other value, my new game will be using the old value.

File: ironmon_tracker/Constants.lua [hunk 2]
```diff
@@ -612,6 +612,28 @@ Constants.PixelImages = {
...
+		{1,0,0,0,0,0,0,0,0},
+		{1,1,0,0,0,0,0,0,0},
+		{1,1,1,0,0,0,0,0,0},
+		{1,1,1,1,0,0,0,0,0},
+		{1,1,1,1,1,0,0,0,0},
+		{1,1,1,1,0,0,0,0,0},
+		{1,1,1,0,0,0,0,0,0},
+		{1,1,0,0,0,0,0,0,0},
+		{1,0,0,0,0,0,0,0,0},
+	},
```
[the remark this answers is not in the document]
Comment (jtigues) [reply ironmon_tracker/Constants.lua:636]: Done. Yeah I figured everything else was a square so it was just a convention

File: ironmon_tracker/Utils.lua [hunk 3]
```diff
@@ -40,6 +40,21 @@ function Utils.getbits(value, startIndex, numBits)
 	return math.floor(Utils.bit_rshift(value, startIndex) % Utils.bit_lshift(1, numBits))
 end
 
+function Utils.generatebitwisemap(value, size)
```
[the remark this answers is not in the document]
Comment (jtigues) [reply ironmon_tracker/Utils.lua:43]: Done

File: ironmon_tracker/Languages/English.lua [hunk 4]
```diff
@@ -563,6 +563,88 @@ ScreenResources{
 		FilterTrainerName = "Trainer Name",
 		FilterRouteName = "Route Name",
 	},
+	BattleEffectsScreen = {
+		Title = "BATTLE EFFECTS",
```
[the remark this answers is not in the document]
Comment (jtigues) [reply ironmon_tracker/Languages/English.lua:567]: Done

File: ironmon_tracker/screens/TrackerScreen.lua [hunk 5]
```diff
@@ -324,6 +324,18 @@ TrackerScreen.Buttons = {
 			Program.redraw(true)
 		end
 	},
+	BattleDetails = {
+		type = Constants.ButtonTypes.NO_BORDER,
+		text = "",
```
[the remark this answers is not in the document]
Comment (jtigues) [reply ironmon_tracker/screens/TrackerScreen.lua:329]: Done

File: ironmon_tracker/Utils.lua [hunk 6]
```diff
@@ -40,6 +40,21 @@ function Utils.getbits(value, startIndex, numBits)
...
+	local map = {}
+	for i=0,size-1, 1 do
+		if value % 2 == 1 then
+			map[i] = true
+		else
+			map[i] = false
+		end
```
[the remark this answers is not in the document]
Comment (jtigues) [reply ironmon_tracker/Utils.lua:50]: Done

**Comment by jtigues** [comment]
expYield has been added back in, and the various concerns with the syntax/formatting have been resolved.

Also added the text strings to the non-English files, oops.

**Review by UTDZac — APPROVED** [review]

**Review by Fellshadow — APPROVED** [review]
Nothing major from me, just the little bits of cleanup stuff I mentioned in the discord
Will leave open since Zac is wanting to do some testing

**Pull request merged**